roachtest: follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum failed

roachtest.follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4407748&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4407748&tab=artifacts#/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum) on master @ [91dfbc8e740ba5792834534dd41af6fb85cee721](https://github.com/cockroachdb/cockroach/commits/91dfbc8e740ba5792834534dd41af6fb85cee721):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum/run_1
	follower_reads.go:712,follower_reads.go:373,follower_reads.go:72,test_runner.go:779: Post "http://34.83.203.224:26258/ts/query": EOF

	test_runner.go:909,test_runner.go:808: test timed out (0s)
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.follower-reads/survival=region/locality=regional/reads=bounded-staleness [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445360&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445360&tab=artifacts#/follower-reads/survival=region/locality=regional/reads=bounded-staleness) on master @ [a7449866f3229872d92e1de973e209838f07de35](https://github.com/cockroachdb/cockroach/commits/a7449866f3229872d92e1de973e209838f07de35):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/follower-reads/survival=region/locality=regional/reads=bounded-staleness/run_1
	follower_reads.go:289,follower_reads.go:72,test_runner.go:779: error verifying node values: pq: internal error: expected bounded_staleness set with a max_timestamp_bound
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #74793 roachtest: follower-reads/survival=region/locality=regional/reads=bounded-staleness failed [C-test-failure O-roachtest O-robot S-1 T-kv branch-release-21.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/survival=region/locality=regional/reads=bounded-staleness.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445360&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4445360&tab=artifacts#/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum) on master @ [a7449866f3229872d92e1de973e209838f07de35](https://github.com/cockroachdb/cockroach/commits/a7449866f3229872d92e1de973e209838f07de35):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum/run_1
	follower_reads.go:289,follower_reads.go:72,test_runner.go:779: error verifying node values: pq: internal error: expected bounded_staleness set with a max_timestamp_bound
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4455314&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4455314&tab=artifacts#/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum) on master @ [978744dbb94b03497154547ce7fa97ea3be9a401](https://github.com/cockroachdb/cockroach/commits/978744dbb94b03497154547ce7fa97ea3be9a401):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum/run_1
	follower_reads.go:289,follower_reads.go:72,test_runner.go:779: error verifying node values: pq: internal error: expected bounded_staleness set with a max_timestamp_bound
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/survival=region/locality=regional/reads=bounded-staleness/insufficient-quorum.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

roachtest.follower-reads/survival=region/locality=regional/reads=bounded-staleness [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4455314&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4455314&tab=artifacts#/follower-reads/survival=region/locality=regional/reads=bounded-staleness) on master @ [978744dbb94b03497154547ce7fa97ea3be9a401](https://github.com/cockroachdb/cockroach/commits/978744dbb94b03497154547ce7fa97ea3be9a401):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /artifacts/follower-reads/survival=region/locality=regional/reads=bounded-staleness/run_1
	follower_reads.go:289,follower_reads.go:72,test_runner.go:779: error verifying node values: pq: internal error: expected bounded_staleness set with a max_timestamp_bound
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #74793 roachtest: follower-reads/survival=region/locality=regional/reads=bounded-staleness failed [C-test-failure O-roachtest O-robot S-1 T-kv branch-release-21.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*follower-reads/survival=region/locality=regional/reads=bounded-staleness.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/boundedstaleness_test.go
@@ -55,16 +55,31 @@ func TestBoundedStalenessEnterpriseLicense(t *testing.T) {
 	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{})
 	defer tc.Stopper().Stop(ctx)
 
-	queries := []string{
-		`SELECT with_max_staleness('10s')`,
-		`SELECT with_min_timestamp(statement_timestamp())`,
+	testCases := []struct {
+		query string
+		args  []interface{}
+	}{
+		{
+			query: `SELECT with_max_staleness('10s')`,
+		},
+		{
+			query: `SELECT with_min_timestamp(statement_timestamp())`,
+		},
+		{
+			query: `SELECT $1::TEXT FROM generate_series(1,1) AS OF SYSTEM TIME with_max_staleness('1s')`,
+			args:  []interface{}{"cat"},
+		},
+		{
+			query: `SELECT $1::TEXT FROM generate_series(1,1) AS OF SYSTEM TIME with_min_timestamp(statement_timestamp())`,
+			args:  []interface{}{"cat"},
+		},
 	}
 
 	defer utilccl.TestingDisableEnterprise()()
 	t.Run("disabled", func(t *testing.T) {
-		for _, q := range queries {
-			t.Run(q, func(t *testing.T) {
-				_, err := tc.Conns[0].QueryContext(ctx, q)
+		for _, testCase := range testCases {
+			t.Run(testCase.query, func(t *testing.T) {
+				_, err := tc.Conns[0].QueryContext(ctx, testCase.query, testCase.args...)
 				require.Error(t, err)
 				require.Contains(t, err.Error(), "use of bounded staleness requires an enterprise license")
 			})
@@ -73,9 +88,9 @@ func TestBoundedStalenessEnterpriseLicense(t *testing.T) {
 
 	t.Run("enabled", func(t *testing.T) {
 		defer utilccl.TestingEnableEnterprise()()
-		for _, q := range queries {
-			t.Run(q, func(t *testing.T) {
-				r, err := tc.Conns[0].QueryContext(ctx, q)
+		for _, testCase := range testCases {
+			t.Run(testCase.query, func(t *testing.T) {
+				r, err := tc.Conns[0].QueryContext(ctx, testCase.query, testCase.args...)
 				require.NoError(t, err)
 				require.NoError(t, r.Close())
 			})

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -756,38 +756,42 @@ func (ex *connExecutor) handleAOST(ctx context.Context, stmt tree.Statement) err
 					return err
 				}
 			}
-		} else if p.extendedEvalCtx.AsOfSystemTime.BoundedStaleness {
-			// This has to be a bounded staleness read with nearest_only=True during
-			// a retry. The AOST read timestamps are expected to differ.
-			if p.extendedEvalCtx.AsOfSystemTime.MaxTimestampBound.IsEmpty() {
-				return errors.AssertionFailedf(
-					"expected bounded_staleness set with a max_timestamp_bound",
-				)
+			return nil
+		}
+		if *p.extendedEvalCtx.AsOfSystemTime == *asOf {
+			// In most cases, the AOST timestamps are expected to match.
+			return nil
+		}
+		if p.extendedEvalCtx.AsOfSystemTime.BoundedStaleness {
+			if !p.extendedEvalCtx.AsOfSystemTime.MaxTimestampBound.IsEmpty() {
+				// This has to be a bounded staleness read with nearest_only=True during
+				// a retry. The AOST read timestamps are expected to differ.
+				return nil
 			}
-		} else if *p.extendedEvalCtx.AsOfSystemTime != *asOf {
-			return errors.AssertionFailedf(
-				"cannot specify AS OF SYSTEM TIME with different timestamps",
-			)
+			return errors.AssertionFailedf("expected bounded_staleness set with a max_timestamp_bound")
 		}
-	} else {
-		// If we're in an explicit txn, we allow AOST but only if it matches with
-		// the transaction's timestamp. This is useful for running AOST statements
-		// using the InternalExecutor inside an external transaction; one might want
-		// to do that to force p.avoidLeasedDescriptors to be set below.
-		if asOf.BoundedStaleness {
-			return pgerror.Newf(
-				pgcode.FeatureNotSupported,
-				"cannot use a bounded staleness query in a transaction",
-			)
-		}
-		if readTs := ex.state.getReadTimestamp(); asOf.Timestamp != readTs {
-			err = pgerror.Newf(pgcode.Syntax,
-				"inconsistent AS OF SYSTEM TIME timestamp; expected: %s", readTs)
-			err = errors.WithHint(err, "try SET TRANSACTION AS OF SYSTEM TIME")
-			return err
-		}
-		p.extendedEvalCtx.AsOfSystemTime = asOf
+		return errors.AssertionFailedf(
+			"cannot specify AS OF SYSTEM TIME with different timestamps. expected: %s, got: %s",
+			p.extendedEvalCtx.AsOfSystemTime.Timestamp, asOf.Timestamp,
+		)
 	}
+	// If we're in an explicit txn, we allow AOST but only if it matches with
+	// the transaction's timestamp. This is useful for running AOST statements
+	// using the InternalExecutor inside an external transaction; one might want
+	// to do that to force p.avoidLeasedDescriptors to be set below.
+	if asOf.BoundedStaleness {
+		return pgerror.Newf(
+			pgcode.FeatureNotSupported,
+			"cannot use a bounded staleness query in a transaction",
+		)
+	}
+	if readTs := ex.state.getReadTimestamp(); asOf.Timestamp != readTs {
+		err = pgerror.Newf(pgcode.Syntax,
+			"inconsistent AS OF SYSTEM TIME timestamp; expected: %s, got: %s", readTs, asOf.Timestamp)
+		err = errors.WithHint(err, "try SET TRANSACTION AS OF SYSTEM TIME")
+		return err
+	}
+	p.extendedEvalCtx.AsOfSystemTime = asOf
 	return nil
 }
 

